cpp/oneapi/dal/algo/pca/backend/gpu/finalize_train_kernel_cov_dpc.cpp

```diff
@@ -102,27 +204,68 @@ static train_result<Task> train(const context_gpu& ctx,
 
     const auto sums =
         pr::table2ndarray_1d<Float>(q, input.get_partial_sum(), sycl::usm::alloc::device);
+    if (desc.get_result_options().test(result_options::means)) {
+        auto [means, means_event] = compute_means(q, rows_count_global, sums, {});
+        means_event.wait_and_throw();
+        result.set_means(homogen_table::wrap(means.flatten(q), 1, column_count));
+    }
+
     const auto xtx =
         pr::table2ndarray<Float>(q, input.get_partial_crossproduct(), sycl::usm::alloc::device);
+    auto [cov, cov_event] = compute_covariance(q, rows_count_global, xtx, sums, {});
+
+    auto [vars, vars_event] = compute_variances(q, cov, { cov_event });
+    vars_event.wait_and_throw();
+    if (desc.get_result_options().test(result_options::vars)) {
+        result.set_variances(homogen_table::wrap(vars.flatten(q), 1, column_count));
+    }
+    auto data_to_compute = cov;
+
+    sycl::event corr_event;
+    if (desc.get_normalization_mode() == normalization::zscore) {
+        auto corr = pr::ndarray<Float, 2>::empty(q, { column_count, column_count }, alloc::device);
+        std::tie(corr, corr_event) =
+            compute_correlation_from_covariance(q, rows_count_global, cov, { cov_event });
```

Comment:
Why do you need to prealocate `corr`?

```suggestion
        pr::ndarray<Float, 2> corr{};
        std::tie(corr, corr_event) =
            compute_correlation_from_covariance(q, rows_count_global, cov, { cov_event });
```